### DISCUSSION TITLE
Potential Architecture Change

### DISCUSSION THREAD

**Discussion by brownoxford** [fenix-hub/godot-engine.jwt#43] [General]
@fenix-hub What are your thoughts on migrating the api to something more similar to [PyJWT](https://pyjwt.readthedocs.io/en/stable/api.html)?

Some notable differences:
- Verification is rolled into decoding and exposed as a single `decode` function
- Internally, code is separated into JWT, JWS, and JWK
- Adheres more closely to [RFC 7519](https://datatracker.ietf.org/doc/html/rfc7519) and [RFC 7515](https://datatracker.ietf.org/doc/html/rfc7515) wrt. separation of duties between an implementation (this library) and an application (a project using this library).

**Comment by fenix-hub** [comment]
Thank you for sharing this library and your thoughts.
I'll make sure to review PyJWT, since I'm not that familiar with python libraries.
I'm ok with considering a migration, even though this will introduce breaking changes for users potentially using this plugin in production environments (at least, it would be a pleasure!).
So I'm willing to leave this thread open to anyone who wants to share their opinion on this topic before making a decision.

Regarding using that library as a reference, I think the most notable point for me would be to actually adhere to the standards, more than how the APIs are exposed and accessed. I would focus on that regardless of how it is done in other solutions. Maybe Python is the closest one to GDScript so in the end it would be a good starting point, but I'd like to keep things "tailored" on Godot and this context, if it makes sense to you.

For example, just from a quick review of PyJWT, I like the idea of having utility functions to actually set standard and custom claims in the token payload more , instead of letting the user put ANY value inside of it. It is true that Json Web Tokens are not *strictly* used for security contexts, so the user should not be limited in what the use case is - but I'd like to give the user some help in avoiding most common errors, since they could really be a pain in a security context.

Anyway any suggestion is really appreciated. What do you think? Are you actively using this library? Is there any particular concern?
